### PR TITLE
Re-name hash inner/byte methods

### PR DESCRIPTION
**Pull request by tcharding**
Currently we have an associated type on hash types `Inner` with accompanying methods `into_inner`, `from_inner`, `as_inner`. Also, we provide a way to create new wrapped hash types. The use of 'inner' becomes ambiguous with the addition of wrapped types because the inner could be the inner hash type or the `Inner` byte array of the inner wrapped hash type.
    
In an effort to make the API more clear and uniform do the following:
    
- Rename `Inner` -> `Bytes`
- Rename `*_inner` -> `*_byte_array`
- Rename the inner hash to/from methods to `*_raw_hash`
    
Correct method prefix `into_` -> `to_` because theses methods convert owned `Copy` types.
    
Add the trait Bound `Copy` to the `Bytes` type because we rely on this trait bound for the conversion methods to be correctly named according to convention.
    
Because of the dependency hole created by `secp256k1` this patch changes the secp dependency to a git tag dependency that includes changes to the hashes calls required so that we can get green lights on CI in this repo.

Fix: #1554

### CODE REVIEW

**Comment by Kixunil** [comment]
Yeah, I think dropping the reference makes sense.

**Comment by tcharding** [comment]
> Yeah, I think dropping the reference makes sense.

I didn't understand this comment so I went digging to try to and realized that I had made the methods non-uniform with the ones for `Inner` hash types. I've re-done the PR so that the wrapper methods `into_raw_hash`, `as_raw_hash`, `from_raw_hash` match the naming and borrowing of the `into_inner`, `as_inner`, `from_inner` methods. Re-wrote the PR description accordingly.

**Review by Kixunil — COMMENTED** [review]
Does it make sense to change the name if we should *also* rename `into_inner` to `to_byte_array`?

**Comment by tcharding** [comment]
`to_inner_hash` and `to_inner_byte_array`?

**Comment by Kixunil** [comment]
Why `inner_byte_array` instead of `byte_array`?

**Comment by tcharding** [comment]
Coolio, I've used `*_raw_hash` and `*_byte_array`. Please see https://github.com/rust-bitcoin/rust-bitcoin/issues/1553 for the solution to the dependency hole problem (DHP) we have that I developed while working on this PR. Converting to draft because we need to resolve the DHP before this can go in. Also this can wait till after your hash work @Kixunil

**Comment by tcharding** [comment]
Note to self, deprecate the old methods.

File: bitcoin/src/address.rs [hunk 1]
```diff
@@ -434,13 +434,13 @@ impl Payload {
     /// Constructs a [Payload] from an output script (`scriptPubkey`).
     pub fn from_script(script: &script::Script) -> Result<Payload, Error> {
         Ok(if script.is_p2pkh() {
-            let mut hash_inner = [0u8; 20];
-            hash_inner.copy_from_slice(&script.as_bytes()[3..23]);
-            Payload::PubkeyHash(PubkeyHash::from_inner(hash_inner))
+            let mut bytes = [0u8; 20];
+            bytes.copy_from_slice(&script.as_bytes()[3..23]);
+            Payload::PubkeyHash(PubkeyHash::from_byte_array(bytes))
```
Comment (Kixunil) [inline bitcoin/src/address.rs:439]: I believe this is simpler:

```suggestion
            let bytes = script.as_bytes()[3..23].try_into().expect("statically 20B long");
            Payload::PubkeyHash(PubkeyHash::from_byte_array(bytes))
```

File: bitcoin/src/address.rs [hunk 2]
```diff
@@ -434,13 +434,13 @@ impl Payload {
...
         } else if script.is_p2sh() {
-            let mut hash_inner = [0u8; 20];
-            hash_inner.copy_from_slice(&script.as_bytes()[2..22]);
-            Payload::ScriptHash(ScriptHash::from_inner(hash_inner))
+            let mut bytes = [0u8; 20];
+            bytes.copy_from_slice(&script.as_bytes()[2..22]);
+            Payload::ScriptHash(ScriptHash::from_byte_array(bytes))
```
Comment (Kixunil) [inline bitcoin/src/address.rs:443]:
```suggestion
            let bytes = script.as_bytes()[2..22].try_into().expect("statically 20B long");
            Payload::ScriptHash(ScriptHash::from_byte_array(bytes))
```

File: hashes/src/lib.rs [hunk 3]
```diff
@@ -205,13 +205,13 @@ pub trait Hash: Copy + Clone + PartialEq + Eq + PartialOrd + Ord +
     const DISPLAY_BACKWARD: bool = false;
 
     /// Unwraps the hash and returns the underlying byte array.
-    fn into_inner(self) -> Self::Inner;
+    fn to_byte_array(self) -> Self::Bytes;
 
     /// Unwraps the hash and returns a reference to the underlying byte array.
```
Comment (Kixunil) [inline hashes/src/lib.rs:210]: Unrelated, but "Unwraps" sounds scary. I'd just write "Returns a reference to the underlying byte array". Same above.

**Review by Kixunil — COMMENTED** [review]
Looks good, will ack once the repo link is updated to the main repo.

File: hashes/src/lib.rs (hunk 3 above)
Comment (tcharding) [reply hashes/src/lib.rs:210]: Sweet, I'll change it while we are here.

File: bitcoin/src/address.rs (hunk 1 above)
Comment (tcharding) [reply bitcoin/src/address.rs]: Used suggestion, as well as below. Thanks

**Comment by tcharding** [comment]
> Looks good, will ack once the repo link is updated to the main repo.

Sweet, I think I've got all green lights now. So after @apoelstra creates a branch on secp I can change the git/branch dep here. FTR all this dependency stuff was painful as f***

**Comment by tcharding** [comment]
> Looks good, will ack once the repo link is updated to the main repo.

Sweet, I think I've got all green lights now. So after @apoelstra creates a branch on secp I can change the git/branch dep here. FTR all this dependency stuff was painful as f***, the quicker we stabalise `hashes` the better. And forget about any outside contributors patching `hashes` its way to involved.

File: Cargo.toml [hunk 4]
```diff
@@ -2,4 +2,5 @@
 members = ["bitcoin", "hashes", "internals"]
 
 [patch.crates-io.bitcoin_hashes]
-path = "hashes"
+git = "https://github.com/tcharding/rust-bitcoin"
+branch = "01-20-rename-hashes-methods"
```
Comment (apoelstra) [inline Cargo.toml:6]: Why is this necessary?

File: bitcoin/Cargo.toml [hunk 5]
```diff
@@ -36,10 +36,10 @@ features = [ "std", "secp-recovery", "base64", "rand", "serde", "bitcoinconsensu
 rustdoc-args = ["--cfg", "docsrs"]
 
 [dependencies]
-bitcoin-internals = { path = "../internals" }
+bitcoin-internals = { git = "https://github.com/tcharding/rust-bitcoin", branch = "01-20-rename-hashes-methods" }
 bech32 = { version = "0.9.0", default-features = false }
 bitcoin_hashes = { version = "0.11.0", default-features = false }
-secp256k1 = { version = "0.25.0", default-features = false, features = ["bitcoin_hashes"] }
+secp256k1 = { git = "https://github.com/tcharding/rust-secp256k1", branch = "01-31-rename-hashes-methods", default-features = false, features = ["bitcoin_hashes"] }
```
Comment (apoelstra) [inline bitcoin/Cargo.toml:42]: This looks fine, though I'd use a tag rather than a branch because you might want to clean up your list of branches one day, but tags probably you'll never delete.

Comment (tcharding) [reply bitcoin/Cargo.toml:42]: Sweet, tag it is. Can you post it please once you've done it so I can use it here (I can't push to `rust-bitcoin/rust-secp256k1`)

File: Cargo.toml (hunk 4 above)
Comment (tcharding) [reply Cargo.toml:6]: Messy isn't it. This is required because otherwise the secp dependency uses a different version of `hashes`. Even though the version _is_ this branch we are on the `path` dependency gives a different version.

Comment (tcharding) [reply Cargo.toml:6]: This is the problem I mention [over here](https://github.com/rust-bitcoin/rust-bitcoin/issues/1553#issuecomment-1410017923). Its caused by the fact that `path` dependency does not patch over git/branch dependency (it only patches over crates.io) so there is, as far as I can telll, no way to get the same version of `hashes` if we use git/branch deps (or tag as you suggest). Am I missing some other way? This really borks things because we can now no longer use `path` dependency, even once this merges, we have to release secp in order to start using `path` again. We can switch to using `master` (or a tagged commit on master) but then we have to update that continually.

Comment (apoelstra) [reply Cargo.toml:6]: Why are you changing the `bitcoin_hashes` dep in the rust-secp branch at all? Just leave it pointing to crates, then the patch should do the right thing.

Comment (tcharding) [reply Cargo.toml]: Bother, I'm on a different machine and I forgot to pull before I force pushed. Have to re-implement the review suggestions from before - won't be long :)

**Comment by tcharding** [comment]
Oh boy, can't believe I overthought that so much :( Oh well, got there in the end. Still needs a branch on `rust-secp25k1` please @apoelstra then I'll update this to us it.

File: bitcoin/Cargo.toml (hunk 5 above)
Comment (apoelstra) [reply bitcoin/Cargo.toml:42]: Pushed. Tag is `2023-02-01--fix-hashes`.

**Comment by tcharding** [comment]
Force push makes a few changes:
- Use git tag dependency for secp
- Change some calls to the new methods to use the least amount of allocations i.e., use `as_` versions if we only need a readable reference
- Add `Copy` trait bound to `Bytes` (see rewritten PR description)

**Comment by apoelstra** [comment]
Wonderful, this looks much simpler!

**Comment by apoelstra** [comment]
Your new use of `flat_map` does not compile with 1.41.0 now. CI is going to fail.

**Comment by tcharding** [comment]
> Your new use of flat_map does not compile with 1.41.0 now. 

Ha, you saw that! Its fixed now :)

**Review by apoelstra — APPROVED** [review]
ACK 5f13d7d9b5c757eef9ac36bb9095d1560535c00f

**Review by Kixunil — APPROVED** [review]
ACK 5f13d7d9b5c757eef9ac36bb9095d1560535c00f

File: bitcoin/examples/taproot-psbt.rs [hunk 6]
```diff
@@ -745,7 +745,7 @@ fn sign_psbt_schnorr(
         Some(_) => keypair, // no tweak for script spend
     };
 
-    let sig = secp.sign_schnorr(&Message::from_slice(&hash.into_inner()[..]).unwrap(), &keypair);
+    let sig = secp.sign_schnorr(&Message::from_slice(&hash.to_byte_array()[..]).unwrap(), &keypair);
```
[the remark this answers is not in the document]
Comment (apoelstra) [reply bitcoin/examples/taproot-psbt.rs:748]: Actually can't we use the `ThirtyTwoByteHash` trait to avoid `Message::from_slice` entirely?

[the remark this answers is not in the document]
Comment (tcharding) [reply bitcoin/examples/taproot-psbt.rs]: Sweet. Rebased and put a new patch at the front using this suggestion.

[the remark this answers is not in the document]
Comment (Kixunil) [reply bitcoin/examples/taproot-psbt.rs]: @apoelstra that's exactly what I suggest.

File: bitcoin/examples/taproot-psbt.rs [hunk 7]
```diff
@@ -745,7 +744,7 @@ fn sign_psbt_schnorr(
         Some(_) => keypair, // no tweak for script spend
     };
 
-    let sig = secp.sign_schnorr(&Message::from_slice(&hash.into_inner()[..]).unwrap(), &keypair);
+    let sig = secp.sign_schnorr(&Message::from(hash), &keypair);
```
Comment (Kixunil) [inline bitcoin/examples/taproot-psbt.rs:747]: `.into()` is shorter but I don't mind.

**Review by Kixunil — APPROVED** [review]
ACK f04cb8360286a800948a99909b0282d679873658

File: bitcoin/examples/taproot-psbt.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (tcharding) [reply bitcoin/examples/taproot-psbt.rs]: Oh, my bad, I was being slow here when I read/implemented the suggestions above. It took me a while to work out how to use `Message::from` and by then I didn't think through your suggestion @Kixunil. I've used `into` now.

**Review by apoelstra — DISMISSED** [review]
ACK 757e5f28ac0eb9ab22128a2555bef55db69cd39d

**Comment by tcharding** [comment]
Rebase only, no other changes.

**Review by apoelstra — DISMISSED** [review]
ACK d05af0f74049673082be63b945910844d76417be

**Review by Kixunil — DISMISSED** [review]
ACK d05af0f74049673082be63b945910844d76417be

**Comment by apoelstra** [comment]
Needs rebase after i#1593, sorry!

**Comment by tcharding** [comment]
Re-based and went over each changed line to check if we were using the optimal call `as_` vs `to_`, changing some of them.

**Comment by tcharding** [comment]
Please let #1190 go in first.

**Review by apoelstra — DISMISSED** [review]
ACK 45bc21edc16df1bd2032c3e0b2888bc2878f939a

**Comment by tcharding** [comment]
Rebased. Can you check I got the changes to `impl_asref_push_bytes` correct please @Kixunil, it looks a bit weird seeing `into()` in the `AsRef` impl.

File: bitcoin/src/hash_types.rs [hunk 8]
```diff
@@ -34,14 +34,14 @@ macro_rules! impl_asref_push_bytes {
             impl AsRef<$crate::blockdata::script::PushBytes> for $hashtype {
                 fn as_ref(&self) -> &$crate::blockdata::script::PushBytes {
                     use $crate::hashes::Hash;
-                    self.as_inner().as_ref()
+                    self.as_byte_array().into()
```
Comment (Kixunil) [inline bitcoin/src/hash_types.rs:37]: I think both should work, while it may look a bit off, it's not wrong.

**Review by Kixunil — APPROVED** [review]
ACK c909c70999a3ed22c7af61bc66d618d5feefdcaa

**Comment by apoelstra** [comment]
Needs rebase after #1190.

File: bitcoin/Cargo.toml [hunk 9]
```diff
@@ -50,7 +50,7 @@ hex_lit = "0.1.1"
 serde_json = "1.0.0"
 serde_test = "1.0.19"
 serde_derive = "1.0.103"
-secp256k1 = { version = "0.25.0", features = ["recovery"] }
+secp256k1 = { git = "https://github.com/rust-bitcoin/rust-secp256k1", tag = "2023-02-01--fix-hashes", features = ["recovery"] }
```
[the remark this answers is not in the document]
Comment (tcharding) [reply bitcoin/Cargo.toml:53]: I couldn't get it to work. Tried again just now and still no dice, does it matter? Here is what I used in case the problem is something obvious
```
[patch.crates-io.secp256k1]
gitt = "https://github.com/rust-bitcoin/rust-secp256k1"
tag = "2023-02-01--fix-hashes"
```

And the error
```
warning: Patch `secp256k1 v0.26.0 (https://github.com/rust-bitcoin/rust-secp256k1?tag=2023-02-01--fix-hashes#b38b94b7)` was not used in the crate graph.
Check that the patched package version and available features are compatible
with the dependency requirements. If the patch has a different version from
what is locked in the Cargo.lock file, run `cargo update` to use the new
version. This may also occur with an optional dependency that is not enabled.
```

**Review by apoelstra — APPROVED** [review]
ACK 0a0b23c90bd9b9fe6389e75a9180e0283a09e48d

File: bitcoin/Cargo.toml (hunk 9 above)
[the remark this answers is not in the document]
Comment (Kixunil) [reply bitcoin/Cargo.toml:53]: The reason is that releasing would be easier - we could keep the patch in and just bump the version. Although now that I think about it, maybe not.

**Comment by tcharding** [comment]
Rebase only, no other changes.

**Comment by apoelstra** [comment]
Heads up that syn 1.0.108 (released 15 hours ago) is going to cause us CI breakage. Need to add another cargo pin.

**Review by apoelstra — APPROVED** [review]
ACK e54192097e6990c83077e8fc02179d3f4734b7ad

**Comment by tcharding** [comment]
Rebase only, no other changes.

**Review by Kixunil — APPROVED** [review]
ACK 161273b209d993db3b9398dac20c5dfe11f61cfd

**Comment by Kixunil** [comment]
Bumped priority - if we can get this in sooner, it'll avoid complex rebases and get us closer to release.
